// @dendronhq/common-test-utils/lib/presets/engine-server/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENGINE_CONFIG_PRESETS = exports.ENGINE_WRITE_PRESETS = exports.ENGINE_QUERY_PRESETS = exports.ENGINE_RENAME_PRESETS = exports.ENGINE_PRESETS_MULTI = exports.ENGINE_PRESETS = exports.ENGINE_SERVER = void 0;
const config_1 = require("./config");
Object.defineProperty(exports, "ENGINE_CONFIG_PRESETS", { enumerable: true, get: function () { return config_1.ENGINE_CONFIG_PRESETS; } });
const delete_1 = require("./delete");
const getByPath_1 = require("./getByPath");
const init_1 = require("./init");
const note_refs_1 = __importDefault(require("./note-refs"));
const query_1 = require("./query");
Object.defineProperty(exports, "ENGINE_QUERY_PRESETS", { enumerable: true, get: function () { return query_1.ENGINE_QUERY_PRESETS; } });
const rename_1 = require("./rename");
Object.defineProperty(exports, "ENGINE_RENAME_PRESETS", { enumerable: true, get: function () { return rename_1.ENGINE_RENAME_PRESETS; } });
const update_1 = require("./update");
var utils_1 = require("./utils");
Object.defineProperty(exports, "ENGINE_HOOKS", { enumerable: true, get: function () { return utils_1.ENGINE_HOOKS; } });
Object.defineProperty(exports, "ENGINE_HOOKS_MULTI", { enumerable: true, get: function () { return utils_1.ENGINE_HOOKS_MULTI; } });
const write_1 = require("./write");
Object.defineProperty(exports, "ENGINE_WRITE_PRESETS", { enumerable: true, get: function () { return write_1.ENGINE_WRITE_PRESETS; } });
exports.ENGINE_SERVER = {
    NOTE_REF: note_refs_1.default,
    ENGINE_WRITE_PRESETS: write_1.ENGINE_WRITE_PRESETS,
    ENGINE_INIT_PRESETS: init_1.ENGINE_INIT_PRESETS,
    ENGINE_UPDATE_PRESETS: update_1.ENGINE_UPDATE_PRESETS,
    ENGINE_DELETE_PRESETS: delete_1.ENGINE_DELETE_PRESETS,
    ENGINE_GET_NOTE_BY_PATH_PRESETS: getByPath_1.ENGINE_GET_NOTE_BY_PATH_PRESETS,
    ENGINE_RENAME_PRESETS: rename_1.ENGINE_RENAME_PRESETS,
    ENGINE_QUERY_PRESETS: query_1.ENGINE_QUERY_PRESETS,
};
exports.ENGINE_PRESETS = [
    { name: "init", presets: exports.ENGINE_SERVER.ENGINE_INIT_PRESETS },
    { name: "delete", presets: exports.ENGINE_SERVER.ENGINE_DELETE_PRESETS },
    { name: "getByPath", presets: exports.ENGINE_SERVER.ENGINE_GET_NOTE_BY_PATH_PRESETS },
    { name: "query", presets: exports.ENGINE_SERVER.ENGINE_QUERY_PRESETS },
    { name: "rename", presets: exports.ENGINE_SERVER.ENGINE_RENAME_PRESETS },
    { name: "update", presets: exports.ENGINE_SERVER.ENGINE_UPDATE_PRESETS },
    { name: "write", presets: exports.ENGINE_SERVER.ENGINE_WRITE_PRESETS },
];
exports.ENGINE_PRESETS_MULTI = [
    { name: "init", presets: exports.ENGINE_SERVER.ENGINE_INIT_PRESETS },
    { name: "delete", presets: exports.ENGINE_SERVER.ENGINE_DELETE_PRESETS },
    { name: "getByPath", presets: exports.ENGINE_SERVER.ENGINE_GET_NOTE_BY_PATH_PRESETS },
    { name: "query", presets: exports.ENGINE_SERVER.ENGINE_QUERY_PRESETS },
    { name: "rename", presets: exports.ENGINE_SERVER.ENGINE_RENAME_PRESETS },
    { name: "update", presets: exports.ENGINE_SERVER.ENGINE_UPDATE_PRESETS },
    { name: "write", presets: write_1.ENGINE_WRITE_PRESETS_MULTI },
];
