// @dendronhq/engine-server/lib/history.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HistoryService = void 0;
const lodash_1 = __importDefault(require("lodash"));
let _HISTORY_SERVICE = undefined;
class HistoryService {
    constructor() {
        this.events = [];
        this.subscribers = {
            engine: [],
            src: [],
            extension: [],
            lspServer: [],
            apiServer: [],
            watcher: [],
        };
        this.pause = false;
    }
    static instance() {
        if (lodash_1.default.isUndefined(_HISTORY_SERVICE)) {
            _HISTORY_SERVICE = new HistoryService();
        }
        return _HISTORY_SERVICE;
    }
    add(event) {
        if (!this.pause) {
            this.events.unshift(event);
            this.subscribers[event.source].forEach((f) => f(event));
        }
    }
    clearSubscriptions() {
        this.subscribers = {
            engine: [],
            src: [],
            extension: [],
            lspServer: [],
            apiServer: [],
            watcher: [],
        };
    }
    lookBack(num = 3) {
        return this.events.slice(0, num);
    }
    subscribe(source, func) {
        this.subscribers[source].push(func);
    }
}
exports.HistoryService = HistoryService;
