// @dendronhq/api-server/lib/modules/notes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoteController = void 0;
const common_all_1 = require("@dendronhq/common-all");
const core_1 = require("../../core");
const memoryStore_1 = require("../../store/memoryStore");
const utils_1 = require("../../utils");
class NoteController {
    static instance() {
        if (!NoteController.singleton) {
            NoteController.singleton = new NoteController();
        }
        return NoteController.singleton;
    }
    async delete({ ws, id, opts, }) {
        const engine = await utils_1.getWS({ ws });
        try {
            const data = await engine.deleteNote(id, opts);
            return data;
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ msg: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
    async getByPath({ ws, ...opts }) {
        const engine = await utils_1.getWS({ ws });
        try {
            const data = await engine.getNoteByPath(opts);
            return data;
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ msg: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
    async query({ ws, ...opts }) {
        const engine = ws
            ? await utils_1.getWS({ ws })
            : memoryStore_1.MemoryStore.instance().getEngine();
        try {
            const data = await engine.queryNotes(opts);
            return data;
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ msg: JSON.stringify(err) }),
                data: [],
            };
        }
    }
    async rename({ ws, ...opts }) {
        const engine = await utils_1.getWS({ ws });
        const ctx = "NoteController:rename";
        try {
            core_1.getLogger().info({ ctx, msg: "enter" });
            const data = await engine.renameNote(opts);
            return data;
        }
        catch (err) {
            core_1.getLogger().error({ ctx, err });
            return {
                error: new common_all_1.DendronError({ payload: err }),
                data: undefined,
            };
        }
    }
    async update({ ws, note, opts, }) {
        const engine = await utils_1.getWS({ ws });
        try {
            await engine.updateNote(note, opts);
            return { error: null };
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ msg: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
}
exports.NoteController = NoteController;
