// @dendronhq/common-test-utils/lib/noteUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoteTestUtilsV4 = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
class NoteTestUtilsV4 {
    static async modifyNoteByPath(opts, cb) {
        const { fname, vault, wsRoot } = opts;
        const npath = path_1.default.join(common_server_1.vault2Path({ vault, wsRoot }), fname + ".md");
        const note = common_server_1.file2Note(npath, vault);
        const newNote = cb(note);
        return await common_server_1.note2File({ note: newNote, vault, wsRoot });
    }
}
exports.NoteTestUtilsV4 = NoteTestUtilsV4;
NoteTestUtilsV4.createSchema = async (opts) => {
    const { fname, vault, noWrite, wsRoot } = lodash_1.default.defaults(opts, {
        noWrite: false,
    });
    let schema = common_all_1.SchemaUtilsV2.createModuleProps({ fname, vault });
    if (opts.modifier) {
        schema = opts.modifier(schema);
    }
    if (!noWrite) {
        const vpath = common_server_1.resolvePath(vault.fsPath, wsRoot);
        await common_server_1.schemaModuleProps2File(schema, vpath, fname);
    }
    return schema;
};
NoteTestUtilsV4.createNote = async (opts) => {
    const { fname, vault, props, body, genRandomId, noWrite, wsRoot, } = lodash_1.default.defaults(opts, { noWrite: false });
    /**
     * Make sure snapshots stay consistent
     */
    const defaultOpts = {
        created: "1",
        updated: "1",
        id: genRandomId ? common_all_1.genUUID() : fname,
    };
    const note = common_all_1.NoteUtilsV2.create({
        ...defaultOpts,
        ...props,
        fname,
        vault,
        body,
    });
    if (!noWrite) {
        await common_server_1.note2File({ note, vault, wsRoot });
    }
    return note;
};
