// @dendronhq/dendron-cli/lib/src/commands/plantSeed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlantSeedCommand = void 0;
const path_1 = __importDefault(require("path"));
const soil_1 = require("./soil");
class PlantSeedCommand extends soil_1.SoilCommand {
    buildArgs(args) {
        super.buildArgs(args);
        args.option("id", {
            describe: "id of seed",
            demandOption: true,
        });
    }
    enrichArgs(args) {
        const opts = super._enrichArgs(args);
        return { ...opts, roots: [args.vault], id: args.id };
    }
    static buildCmd(yargs) {
        const _cmd = new PlantSeedCommand();
        return yargs.command("plantSeed", "plant a seed", _cmd.buildArgs, _cmd.eval);
    }
    async execute(opts) {
        const { roots, wsRoot } = opts;
        const node_modules = path_1.default.join(opts.wsRoot, "node_modules");
        const SeedClass = require(path_1.default.join(node_modules, opts.id)).default;
        const seed = new SeedClass({ name: SeedClass.name, roots, wsRoot });
        await seed.plant();
        return;
    }
}
exports.PlantSeedCommand = PlantSeedCommand;
