// @dendronhq/api-server/lib/modules/workspace/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceController = void 0;
const common_all_1 = require("@dendronhq/common-all");
const engine_server_1 = require("@dendronhq/engine-server");
const core_1 = require("../../core");
const utils_1 = require("../../utils");
class WorkspaceController {
    static instance() {
        if (!WorkspaceController.singleton) {
            WorkspaceController.singleton = new WorkspaceController();
        }
        return WorkspaceController.singleton;
    }
    async init({ uri, config }) {
        let notes;
        let schemas;
        const { vaults } = config;
        const logger = core_1.getLogger();
        const engine = engine_server_1.DendronEngineV2.createV3({
            wsRoot: uri,
            vaults: vaults.map((ent) => ({ fsPath: ent })),
            logger,
        });
        const { error } = await engine.init();
        if (error && error.code !== common_all_1.ERROR_CODES.MINOR) {
            error.friendly = "error initializing notes";
            return { error };
        }
        notes = engine.notes;
        schemas = engine.schemas;
        await utils_1.putWS({ ws: uri, engine });
        const payload = {
            error,
            data: { notes, schemas },
        };
        return payload;
    }
    async sync({ ws }) {
        const engine = await utils_1.getWS({ ws });
        const { notes, schemas } = engine;
        return {
            error: null,
            data: { notes, schemas },
        };
    }
}
exports.WorkspaceController = WorkspaceController;
