// @dendronhq/common-server/lib/parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaParserV2 = exports.ParserBaseV2 = void 0;
const common_all_1 = require("@dendronhq/common-all");
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const filesv2_1 = require("./filesv2");
class ParserBaseV2 {
    constructor(opts) {
        this.opts = opts;
    }
    get logger() {
        return this.opts.logger;
    }
}
exports.ParserBaseV2 = ParserBaseV2;
class SchemaParserV2 extends ParserBaseV2 {
    static parseRaw(schemaOpts, opts) {
        const version = lodash_1.default.isArray(schemaOpts) ? 0 : 1;
        if (version > 0) {
            return SchemaParserV2.parseSchemaModuleOpts(schemaOpts, opts);
        }
        else {
            // TODO: legacy
            const schemaDict = {};
            schemaOpts.map((ent) => {
                const schema = common_all_1.SchemaUtilsV2.create(ent);
                schemaDict[schema.id] = schema;
            });
            const maybeRoot = lodash_1.default.find(lodash_1.default.values(schemaDict), {
                parent: "root",
            });
            return {
                version: 0,
                root: maybeRoot,
                schemas: schemaDict,
                fname: opts.fname,
                vault: opts.root,
            };
        }
    }
    static parseSchemaModuleOpts(schemaModuleProps, opts) {
        const { imports, schemas, version } = schemaModuleProps;
        const { fname, root } = opts;
        let schemaModulesFromImport = lodash_1.default.flatMap(imports, (ent) => {
            const fpath = path_1.default.join(root.fsPath, ent + ".schema.yml");
            return filesv2_1.file2Schema(fpath);
        });
        const schemaPropsFromImport = schemaModulesFromImport.flatMap((mod) => {
            const domain = mod.fname;
            return lodash_1.default.values(mod.schemas).map((ent) => {
                ent.data.pattern = ent.data.pattern || ent.id;
                ent.id = `${domain}.${ent.id}`;
                ent.fname = fname;
                ent.parent = null;
                ent.children = ent.children.map((ent) => `${domain}.${ent}`);
                ent.vault = root;
                return ent;
            });
        });
        const schemaPropsFromFile = schemas.map((ent) => {
            return common_all_1.SchemaUtilsV2.create({ ...ent, vault: root });
        });
        const schemasAll = schemaPropsFromImport.concat(schemaPropsFromFile);
        const schemasDict = {};
        schemasAll.forEach((ent) => {
            schemasDict[ent.id] = ent;
        });
        const rootModule = common_all_1.SchemaUtilsV2.getModuleRoot(schemaModuleProps);
        const addConnections = (parent) => {
            lodash_1.default.map(parent.children, (ch) => {
                const child = schemasDict[ch];
                if (!child) {
                    throw new common_all_1.DendronError({
                        status: common_all_1.ENGINE_ERROR_CODES.MISSING_SCHEMA,
                        msg: JSON.stringify({ parent, missingChild: ch }),
                    });
                }
                common_all_1.DNodeUtilsV2.addChild(parent, child);
                return addConnections(child);
            });
        };
        // add parent relationship
        addConnections(rootModule);
        return {
            version,
            imports,
            root: rootModule,
            schemas: schemasDict,
            fname,
            vault: root,
        };
    }
}
exports.SchemaParserV2 = SchemaParserV2;
